Include/internal/pycore_import.h

```diff
@@ -83,6 +83,12 @@ struct _import_state {
        This is initialized lazily in PyState_AddModule(), which is also
        where modules get added. */
     PyObject *modules_by_index;
+#ifdef Py_GIL_DISABLED
+    /* This list contains GIL slots (see Py_mod_gil) for modules that may be
+       reinitialized by reload_singlephase_extension(). It is indexed the same
+       way as modules_by_index. */
+    PyObject *module_gil_by_index;
+#endif
```

Comment:
Won't caching this per-interpreter be a problem?  It may make more sense storing this on the global extensions cache.

Reply:
Also, what's the advantage to adding a new cache rather than storing the value in `modules_by_index`?

Reply:
> I avoided that because it wouldn't make sense to change what it holds in both build types, and changing the format of the data only in free-threaded builds felt more complicated overall.

That makes sense.  Thanks for explaining.

> My understanding is that because this only matters for the `m_size == -1` case, the relevant modules don't support subinterpreters, and this logic only applies to recreating a module from a stored dict, within the main interpreter.

Legacy (non-isolated) subinterpreters do support single-phase init modules.

Reply:
> Is that the cache used [here](https://github.com/python/cpython/blob/1b639a04cab0e858d90e2ac459fb34b73700701f/Python/import.c#L1318-L1319)?

That's right.

> The same reasoning applies as why I didn't put it in `modules_by_index` - we'd have to choose between either changing the contents of the map in all build types, or having different data formats between build types. I can do one of those if you think it's worth it.

The difference is that the global cache has values of type `struct extensions_cache_value` (in import.c) so adding another value, conditional or not, isn't a big deal.

Reply:
> Legacy (non-isolated) subinterpreters do support single-phase init modules.

I was going based off of [the documentation for PyModuleDef.m_size](https://docs.python.org/3.13/c-api/module.html#c.PyModuleDef.m_size), which says "Setting m_size to -1 means that the module does not support sub-interpreters, because it has global state." Should that instead say something like "...the module does not support *isolated* sub-interpreters..."?

> the global cache has values of type `struct extensions_cache_value`

Aha, looks like that's pretty new, from 291cfa454b. In the version of `main` I was working with from a couple days ago, the cache only held a `PyModuleDef*`. This should be much nicer, thanks.

Reply:
> Should that instead say something like "...the module does not support _isolated_ sub-interpreters..."?

Maybe.  Technically, it does mean the module does not support subinterpreters.  However, that was never enforced for legacy subinterpreters, so we've preserved the existing behavior.  Ultimately, I'd rather we move away from legacy subinterpreters and avoid making a distinction where possible.  Clearly that led to confusion for you here. 😞